File: September/4/longvalidparentheses.py
#O(n)
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        stack = [-1]  # base index
        max_len = 0

        for i, char in enumerate(s):
            if char == '(':
                stack.append(i)
            else:
                stack.pop()
                if not stack:
                    stack.append(i)
                else:
                    max_len = max(max_len, i - stack[-1])

        return max_len
    
    def longestValidParenthesesDP(self, s: str) -> int:
        n = len(s)
        if n == 0:
            return 0

        dp = [0] * n
        max_len = 0

        for i in range(1, n):
            if s[i] == ')':
                if s[i-1] == '(':
                    dp[i] = (dp[i-2] if i >= 2 else 0) + 2
                elif i - dp[i-1] - 1 >= 0 and s[i - dp[i-1] - 1] == '(':
                    dp[i] = dp[i-1] + 2 + (dp[i - dp[i-1] - 2] if i - dp[i-1] >= 2 else 0)
                max_len = max(max_len, dp[i])

        return max_len

File: September/4/longvalidparentheses.php
<?php
class Solution {
    function longestValidParentheses($s) {
        $stack = [-1];  // base index
        $max_len = 0;

        for ($i = 0; $i < strlen($s); $i++) {
            if ($s[$i] === '(') {
                array_push($stack, $i);
            } else {
                array_pop($stack);
                if (empty($stack)) {
                    array_push($stack, $i);
                } else {
                    $max_len = max($max_len, $i - end($stack));
                }
            }
        }

        return $max_len;
    }
}

// Example Usage
$sol = new Solution();
echo $sol->longestValidParentheses("(()") . "\n";       // Output: 2
echo $sol->longestValidParentheses(")()())") . "\n";    // Output: 4
echo $sol->longestValidParentheses("") . "\n";          // Output: 0
?>
